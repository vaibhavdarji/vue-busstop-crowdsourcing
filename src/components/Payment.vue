<template>
  <v-card>
    <v-form ref="form" @submit.prevent="onNext">
      <v-container grid-list-xl fluid>
        <template v-if="errorMsg">
          <custom-error :msg="errorMsg" @onErrorBtnClick="onNext"></custom-error>
        </template>
        <v-layout row>
          <v-flex xs12>
           <h2>
            Your selected amount:  
            <v-icon left>attach_money</v-icon>
            {{userDonation.amount}}
           </h2>
          </v-flex>  
        </v-layout>
        <v-layout row>
          <v-flex xs8>
            <v-text-field 
              label="Card No"
              :rules="rules.cardNumber"
              v-model="form.cardNo"
              hint="1231 - 2312 - 3123 - 1231"
              mask="credit-card"   
              required
            ></v-text-field>
          </v-flex>  
        </v-layout>
         <v-layout row>
           <v-flex xs8>
            <v-text-field 
              label="Name of AC holder"
              hint="Joseph"
              :rules="rules.cardName"
              v-model="form.cardUserName"
              required
            ></v-text-field>
          </v-flex>  
         </v-layout>
        <v-layout row>
          <v-flex xs8>
            <v-text-field 
              :rules="rules.expiryDate"
              mask="##/####"
              hint="MM/YYYY"
              label="Expiry Date"
              v-model="form.cardExp"
              required>
            </v-text-field>
          </v-flex>  
          <v-flex xs4>
            <v-text-field 
              label="CVV"
              hint="1234"
              mask="####"
               :rules="rules.cvc"
              v-model="form.cardCVC"
              required>
            </v-text-field>
          </v-flex> 
          
        </v-layout>
        <v-layout row >
          <v-flex xs12>
            <h4 class="error--text" v-if="transactionError">{{transactionError}}</h4>
          </v-flex> 
        </v-layout>
        <v-card-actions>
          <custom-button @onClick="onPrevious" :disabled="loading">
          Previous 
        </custom-button>
        <v-spacer></v-spacer>
        <custom-button :isLoading="loading" type="submit" :disabled="isBtnDisable">
          Make Payment 
        </custom-button>
      </v-card-actions>
      </v-container>
    </v-form>
  </v-card>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import CustomButton from '@/components/Button.vue';
import CustomError from '@/components/Error.vue';
import {
  UPDATE_USER_DONATION_DETAIL,
  CHARGE_USER_FOR_DONATION,
  UPDATE_BUS_STOP_DONATION
} from '@/actions/types';

export default {
  name: 'Payment',
  data () {
    const defaultForm = Object.freeze({
      cardNo: '',
      cardExp: '',
      cardCVC: '',
      cardUserName: ''
    });
    return {
      errorMsg: '',
      doRefresh: false,
      form: Object.assign({}, defaultForm),
      loading: false,
      loader: null,
      transactionError: null,
      rules: {
        cardNumber: [
          value => {
            return (value || '').trim().length === 16 || 'Enter valid credit card number';
          }
        ],
        cardName: [
          value => {
            return (value || '').trim().length > 0 || 'Enter card holder name';
          }
        ],
        cvc: [
          value => {
            return (value || '').trim().length > 2  || 'Enter valid CVC number';
          }
        ],
        expiryDate: [
          value => {
            return (value || '').trim().length === 6 || 'Enter valid expiry date';
          }
        ]
      },
      defaultForm
    };
  },

  components: {
    CustomButton,
    CustomError
  },

  computed: {
    ...mapGetters([
      'userDonation'
    ]),

    formIsValid () {
      return (this.form.cardNo && this.form.cardUserName && this.form.cardExp && this.form.cardCVC);
    },

    isBtnDisable () {
      return !this.formIsValid || this.loading;
    }
  },

  methods: {
    ...mapActions({
      'updateUserDonationDetail': UPDATE_USER_DONATION_DETAIL,
      'chargeUserForDonation': CHARGE_USER_FOR_DONATION,
      'updateBusStopDonation': UPDATE_BUS_STOP_DONATION
    }),
    onPrevious () {
      this.$emit('onPreviousStep');
    },

    onNext () {
      this.errorMsg = '';
      this.loader = 'loading';
      const l = this.loader;
      this[l] = !this[l];
      this.updateUserDonationDetail(this.form);
      this.chargeUserForDonation().then(() => {
        return this.updateBusStopDonation();
      }).then(() => {
        this[l] = false;
        this.loader = null;
        this.$emit('onNextStep');
      }).catch((error) => {
        this.errorMsg = error.message;
        this[l] = false;
        this.loader = null;
      });
    }
  }
}
</script>

