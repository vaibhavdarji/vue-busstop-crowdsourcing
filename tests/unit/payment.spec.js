import { shallowMount, mount } from "@vue/test-utils";
import Payment from '@/components/Payment.vue';

import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";
import Constants from '@/constants';

Vue.use(Vuex);
Vue.use(Vuetify);

describe("Payment.vue", () => {
  let getters, store;
  beforeEach(() => {
    getters = {
      userDonation: () => ({
        ...Constants.USER_INFO
      })
    };
    store = new Vuex.Store({
      getters
    });
  });

  it("should match snapshot", () => {
    const wrapper = shallowMount(Payment, {
      store,
      Vue
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});