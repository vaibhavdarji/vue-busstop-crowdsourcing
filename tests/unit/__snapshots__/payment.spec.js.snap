// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment.vue should match snapshot 1`] = `
<v-card-stub
  tag="div"
>
  <v-form-stub>
    <v-container-stub
      fluid=""
      grid-list-xl=""
      tag="div"
    >
      <!---->
       
      <v-layout-stub
        row=""
        tag="div"
      >
        <v-flex-stub
          tag="div"
          xs12=""
        >
          <h2>
            
          Your selected amount:  
          
            <v-icon-stub
              left=""
            >
              attach_money
            </v-icon-stub>
            
          0
         
          </h2>
        </v-flex-stub>
      </v-layout-stub>
       
      <v-layout-stub
        row=""
        tag="div"
      >
        <v-flex-stub
          tag="div"
          xs8=""
        >
          <v-text-field-stub
            hint="1231 - 2312 - 3123 - 1231"
            label="Card No"
            mask="credit-card"
            required=""
            rules="value => {
          return (value || '').trim().length === 16 || 'Enter valid credit card number';
        }"
            value=""
          />
        </v-flex-stub>
      </v-layout-stub>
       
      <v-layout-stub
        row=""
        tag="div"
      >
        <v-flex-stub
          tag="div"
          xs8=""
        >
          <v-text-field-stub
            hint="Joseph"
            label="Name of AC holder"
            required=""
            rules="value => {
          return (value || '').trim().length > 0 || 'Enter card holder name';
        }"
            value=""
          />
        </v-flex-stub>
      </v-layout-stub>
       
      <v-layout-stub
        row=""
        tag="div"
      >
        <v-flex-stub
          tag="div"
          xs8=""
        >
          <v-text-field-stub
            hint="MM/YYYY"
            label="Expiry Date"
            mask="##/####"
            required=""
            rules="value => {
          return (value || '').trim().length === 6 || 'Enter valid expiry date';
        }"
            value=""
          />
        </v-flex-stub>
         
        <v-flex-stub
          tag="div"
          xs4=""
        >
          <v-text-field-stub
            hint="1234"
            label="CVV"
            mask="####"
            required=""
            rules="value => {
          return (value || '').trim().length > 2 || 'Enter valid CVC number';
        }"
            value=""
          />
        </v-flex-stub>
      </v-layout-stub>
       
      <v-layout-stub
        row=""
        tag="div"
      >
        <v-flex-stub
          tag="div"
          xs12=""
        >
          <!---->
        </v-flex-stub>
      </v-layout-stub>
       
      <v-card-actions-stub>
        <custom-button-stub
          btntype="button"
        >
          
        Previous 
      
        </custom-button-stub>
         
        <v-spacer-stub />
         
        <custom-button-stub
          btntype="button"
          disabled="true"
          type="submit"
        >
          
        Make Payment 
      
        </custom-button-stub>
      </v-card-actions-stub>
    </v-container-stub>
  </v-form-stub>
</v-card-stub>
`;
